File: low-level-design-java/03_design_patterns/01_creational/03_builder/Computer.java
public class Computer {
    // Required parameters
    private String CPU;
    private String RAM;

    // Optional parameters
    private boolean hasGraphicsCard;
    private boolean hasBluetooth;

    // Private constructor
    private Computer(Builder builder) {
        this.CPU = builder.CPU;
        this.RAM = builder.RAM;
        this.hasGraphicsCard = builder.hasGraphicsCard;
        this.hasBluetooth = builder.hasBluetooth;
    }

    public String toString() {
        return "Computer [CPU=" + CPU + ", RAM=" + RAM +
                ", GraphicsCard=" + hasGraphicsCard + ", Bluetooth=" + hasBluetooth + "]";
    }

    // Static nested Builder class
    public static class Builder {
        // Required
        private String CPU;
        private String RAM;

        // Optional
        private boolean hasGraphicsCard;
        private boolean hasBluetooth;

        public Builder(String CPU, String RAM) {
            this.CPU = CPU;
            this.RAM = RAM;
        }

        public Builder setGraphicsCard(boolean hasGraphicsCard) {
            this.hasGraphicsCard = hasGraphicsCard;
            return this;
        }

        public Builder setBluetooth(boolean hasBluetooth) {
            this.hasBluetooth = hasBluetooth;
            return this;
        }

        public Computer build() {
            return new Computer(this);
        }
    }
}

File: low-level-design-cpp/03_design_patterns/01_creational/03_builder/Computer.cpp
#include <iostream>
#include <string>

class Computer {
private:
    // Required parameters
    std::string CPU;
    std::string RAM;

    // Optional parameters
    bool hasGraphicsCard;
    bool hasBluetooth;

    // Private constructor
    Computer(const std::string& cpu, const std::string& ram, bool graphics, bool bluetooth)
        : CPU(cpu), RAM(ram), hasGraphicsCard(graphics), hasBluetooth(bluetooth) {}

public:
    // Builder Class
    class Builder {
    private:
        std::string CPU;
        std::string RAM;
        bool hasGraphicsCard = false;
        bool hasBluetooth = false;

    public:
        Builder(const std::string& cpu, const std::string& ram)
            : CPU(cpu), RAM(ram) {}

        Builder& setGraphicsCard(bool value) {
            hasGraphicsCard = value;
            return *this;
        }

        Builder& setBluetooth(bool value) {
            hasBluetooth = value;
            return *this;
        }

        Computer build() {
            return Computer(CPU, RAM, hasGraphicsCard, hasBluetooth);
        }
    };

    // Output description
    void printSpecs() const {
        std::cout << "Computer [CPU=" << CPU
                  << ", RAM=" << RAM
                  << ", GraphicsCard=" << (hasGraphicsCard ? "true" : "false")
                  << ", Bluetooth=" << (hasBluetooth ? "true" : "false")
                  << "]" << std::endl;
    }
};